### PR TITLE
Client side changes for anonymous mode

### PR DESCRIPTION
**Pull request by kimjyhello**
Description
-----------
- Give warning on sharing anonymous run link.
- Disallow logging artifacts in anonymous mode.

Screen Shot 2021-10-22 at 12 36 05 PM
Screen Shot 2021-10-22 at 12 37 44 PM

Testing
-------

How was this PR tested?

Release Notes
-------------

Below, please enter user-facing release notes as one or more bullet points.
If your change is not user-visible, write `NO RELEASE NOTES` instead, with no bullet points.

------------- BEGIN RELEASE NOTES ------------------
NO RELEASE NOTES
------------- END RELEASE NOTES --------------------

### CODE REVIEW

File: tests/wandb_run_test.py [hunk 1]
```diff
@@ -300,3 +300,42 @@ def test_unlogged_artifact_in_config(live_mock_server, test_settings):
...
             == "Cannot json encode artifact before it has been logged or in offline mode."
         )
+
+
+def test_anonymous_mode(live_mock_server, test_settings, capsys):
+    api_key = os.environ.pop("WANDB_API_KEY")
+    uname = os.environ.pop("WANDB_USERNAME")
+    test_settings.update({"anonymous": "must", "api_key": None})
+
+    run = wandb.init(settings=test_settings, anonymous="must")
```
Comment (annirudh) [inline tests/wandb_run_test.py:310]: If for some reason there's an exception thrown in `wandb.init`, does that mean that we might not set `WANDB_USERNAME` and `WANDB_API_KEY` back to the original values? That could cause the other tests to fail unexpectedly.

Ideally, we would have something like:

```
with new_env(lambda env: # modify env as needed):
  run = wandb.init(...)
```

So that even if there's an error, the `with` guarantees that we restore the original env.

File: tests/wandb_run_test.py [hunk 2]
```diff
@@ -300,3 +300,42 @@ def test_unlogged_artifact_in_config(live_mock_server, test_settings):
             str(e_info.value)
             == "Cannot json encode artifact before it has been logged or in offline mode."
         )
+
+
+def test_anonymous_mode(live_mock_server, test_settings, capsys):
+    api_key = os.environ.pop("WANDB_API_KEY")
+    uname = os.environ.pop("WANDB_USERNAME")
+    test_settings.update({"anonymous": "must", "api_key": None})
```
Comment (annirudh) [inline tests/wandb_run_test.py:308]: Does this persist after the test? It would be safer to copy the settings instead of updating it in place as otherwise other tests could observe the effects of this update.

File: wandb/sdk/wandb_run.py [hunk 3]
```diff
@@ -44,7 +44,7 @@
 from wandb import errors
 from wandb import trigger
 from wandb._globals import _datatypes_set_callback
-from wandb.apis import internal, public
+from wandb.apis import internal, InternalApi, public
```
Comment (annirudh) [inline wandb/sdk/wandb_run.py:47]: You technically don't need this import. You should be able to do `internal.Api(...)` to init an internal API object.

**Review by annirudh — COMMENTED** [review]
Looks good overall, just had some questions about lingering test state.

File: tests/wandb_run_test.py (hunk 1 above)
Comment (kptkin) [reply tests/wandb_run_test.py:310]: you can mock/patch environ variables:
https://adamj.eu/tech/2020/10/13/how-to-mock-environment-variables-with-pytest/
https://docs.pytest.org/en/stable/monkeypatch.html#monkeypatching-environment-variables

File: tests/wandb_run_test.py (hunk 2 above)
Comment (KyleGoyette) [reply tests/wandb_run_test.py]: These changes don't persist after the test, and its a common pattern in our cli tests.

Comment (annirudh) [reply tests/wandb_run_test.py]: Gotcha, thanks! Is that handled by the pytest fixtures? Trying to understand how the env gets reset each test.

File: wandb/sdk/wandb_run.py [hunk 4]
```diff
@@ -2499,6 +2506,9 @@ def _log_artifact(
         is_user_created: bool = False,
         use_after_commit: bool = False,
     ) -> wandb_artifacts.Artifact:
+        api = internal.Api()
+        if api.settings().get("anonymous") == "true":
+            wandb.termwarn("Artifacts won't be claimed when you claim the run.")
```
Comment (annirudh) [inline wandb/sdk/wandb_run.py:2511]: Maybe reword to: `Artifacts logged anonymously cannot be claimed and expire after 7 days.`

File: tests/wandb_run_test.py (hunk 2 above)
Comment (KyleGoyette) [reply tests/wandb_run_test.py]: Oh, yeah for `test_settings` pytest fixtures handles the resetting of the value. But the environment is not always reset between tests. We've switched over to using a patched environment context as you suggested. Great idea!

**Review by annirudh — APPROVED** [review]
